Fixed the flaky test_rref_context_debug_info

osalpekar opened the pull request:
Summary:
`test_rref_context_debug_info` was flaky since the message from owner RRef confirming fork could rarely arrive after the test checked whether the pending User RRefs map was empty - leading to an assertion error. This diff creates a utility function that should be used by any test to wait for this message to complete processing before doing any assertions related to the pending User RRefs map.

GitHub Issue: https://github.com/pytorch/pytorch/issues/30988

Test Plan: Stress tested `test_rref_context_debug_info` 200 times.

Differential Revision: D19612289

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1303,6 +1306,7 @@ def test_rref_context_debug_info(self):
         rref1 = RRef(self.rank)
 
         # don't need a barrier here as local RRef is handled by this thread
+        wait_until_pending_users_flushed()
```
Do we need this here? As rref1 is the local OwnerRRef, there shouldn't be any user at this point?

osalpekar replied:
Yeah we shouldn't need it here. I added this before all assertions on `num_pending_users` but it's probably better to remove it from here

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -14,7 +14,13 @@
+    wait_until_node_failure,
+    initialize_pg,
+    get_shutdown_error_regex,
+    wait_until_pending_users_flushed,
```
can we sort this based on alphabet order?

mrshenli reviewed (approved):
Thanks a lot for fixing!!

osalpekar committed:
Fixed the flaky test_rref_context_debug_info (#32749)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32749

The test was flaky since the message from owner RRef confirming fork would arrive after the test checked whether the pending User RRefs map was empty - leading to an assertion error. This diff creates a utility function that should be used by any test to wait for this message to complete processing before doing any assertions related to the pending User RRefs map.

GitHub Issue: https://github.com/pytorch/pytorch/issues/30988

Test Plan: Stress tested `test_rref_context_debug_info` 200 times.

Differential Revision: D19612289

fbshipit-source-id: 3dd9f186028e2afc26a1b1ebc18b7e64249977fc

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit d6d5f8f2:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d6d5f8f29651b57bf1d05f15567c7c5f1cd9db82).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4392131) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93897462) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93897462)) :snowflake:
<details>
<summary>
<code>Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_reshape Fatal Python error: Segmentation fault</code>
</summary>

```
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_transpose_matmul PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_unsqueeze PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_error_on_data_parallel PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_is_in_onnx_export PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_strip_doc_string PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_validate_dynamic_axes_invalid_input_output_name PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_concat PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_div PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_lstm PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_mul PASSED [ 99%] 
Jan 31 22:26:52 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_reshape Fatal Python error: Segmentation fault 
Jan 31 22:26:52  
Jan 31 22:26:52 Current thread 0x00007f5e398ab700 (most recent call first): 
Jan 31 22:26:52   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 43 in repr_instance 
Jan 31 22:26:52   File "/usr/lib/python3.6/reprlib.py", line 65 in repr1 
Jan 31 22:26:52   File "/usr/lib/python3.6/reprlib.py", line 55 in repr 
Jan 31 22:26:52   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 36 in repr 
Jan 31 22:26:52   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 67 in saferepr 
Jan 31 22:26:52   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 655 in repr_args 
Jan 31 22:26:52   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 736 in repr_traceback_entry 
Jan 31 22:26:52   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 777 in repr_traceback 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
